File: leetcode/123-best-time-to-buy-and-sell-stock-III/main.go
package main

import (
	"fmt"
	"math"
)

func maxProfit(prices []int) int {
	firstSell, secondSell := 0, 0
	firstBuy, secondBuy := math.MinInt32, math.MinInt32

	for _, price := range prices {
		secondSell = max(secondSell, secondBuy+price)
		secondBuy = max(secondBuy, firstSell-price)
		firstSell = max(firstSell, firstBuy+price)
		firstBuy = max(firstBuy, -price)
	}

	return secondSell
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	prices := []int{3, 3, 5, 0, 0, 3, 1, 4}
	fmt.Println(maxProfit(prices))
}

File: leetcode/123-best-time-to-buy-and-sell-stock-III/main.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        first_sell, second_sell = 0, 0
        first_buy, second_buy = -float('inf'), -float('inf')

        for price in prices:
            second_sell = max(second_sell, second_buy+price)
            second_buy = max(second_buy, first_sell-price)
            first_sell = max(first_sell, first_buy+price)
            first_buy = max(first_buy, -price)

        return second_sell


solution = Solution()
print(solution.maxProfit([3, 3, 5, 0, 0, 3, 1, 4]))
